module/preference/PreferenceDialog.kt

```diff
@@ -0,0 +1,332 @@
+import com.vaticle.typedb.studio.framework.material.Navigator
+import com.vaticle.typedb.studio.framework.material.Navigator.rememberNavigatorState
+import com.vaticle.typedb.studio.framework.material.Separator
+import com.vaticle.typedb.studio.state.StudioState
+import com.vaticle.typedb.studio.state.common.util.Label
+import com.vaticle.typedb.studio.state.page.Navigable
+
+object PreferenceDialog {
+    private val WIDTH = 800.dp
+    private val HEIGHT = 600.dp
+    private val NAVIGATOR_INIT_SIZE = 200.dp
+    private val NAVIGATOR_MIN_SIZE = 150.dp
+    private val STATE_INIT_SIZE = 600.dp
+    private val STATE_MIN_SIZE = 500.dp
+    private val QUERY_LIMIT_PLACEHOLDER = "1000"
+    private val IGNORED_PATHS_PLACEHOLDER = ".git, .typedb-studio"
+
+    private val appData = StudioState.appData.preferences
+    private var focusedPreferenceGroup by mutableStateOf(PreferenceGroup(""))
+
+    sealed interface Preference {
+        @Composable fun display()
+        fun valid(): Boolean
+
+        class TextInput(initial: String, var label: String, private var placeholder: String,
+                        var validator: (String) -> Boolean = { true }) : Preference {
+
+            var value by mutableStateOf(initial)
+
+            override fun valid(): Boolean {
+                return validator(value)
+            }
+
+            @Composable
+            override fun display() {
+                var border = Form.Border(1.dp, RoundedCornerShape(Theme.ROUNDED_CORNER_RADIUS)) {
+                    if (this.valid()) {
+                        Theme.studio.border
+                    } else {
+                        Theme.studio.errorStroke
+                    }
+                }
+
+                Field(label = label) {
+                    TextInput(
+                        value = value,
+                        placeholder = placeholder,
+                        border = border,
+                        onValueChange = { value = it }
+                    )
+                }
+            }
+        }
+
+        class Checkbox(initial: Boolean, var label: String): Preference {
+            var value by mutableStateOf(initial)
+            @Composable
+            override fun display() {
+                Field(label = label) {
+                    Checkbox(
+                        value = value,
+                        onChange = { value = it }
+                    )
+                }
+            }
+
+            override fun valid(): Boolean {
+                return true
+            }
+        }
+    }
+
+
+    class PreferencesForm : State {
+        // Graph Visualiser Preferences
+        var graphOutput = Preference.Checkbox(appData.graphOutput, Label.ENABLE_GRAPH_OUTPUT)
+        // Project Manager Preferences
+        val ignoredPathsData = appData.ignoredPaths.toString()
+        var ignoredPaths = Preference.TextInput(ignoredPathsData.substring(1, ignoredPathsData.length - 1),
```

Comment:
Huh? `.substring`? Why?

Is this because `ignoredPathsData` starts and ends with a quote? If so, we can clarify that with `trim` instead of `substring`, and push it into the line above. Let's not degrade the clarity of what this parameter to `Preference.TextInput` is.

Although, I'd still consider converting this to a named parameter just to be on the safe side.